File: backend/src/schemas/market.py
"""Market-related Pydantic schemas."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class MarketFilters(BaseModel):
    """Query parameters for filtering markets."""

    algorithm: Literal["a", "b"] = "a"
    market_type: str | None = None
    min_value_score: float = Field(default=0, ge=0, le=100)
    min_confidence: float = Field(default=0, ge=0)
    limit: int = Field(default=50, ge=1, le=200)
    offset: int = Field(default=0, ge=0)


class GameInfo(BaseModel):
    """Embedded game information."""

    game_id: str
    home_team: str
    away_team: str
    tip_time: datetime
    status: Literal["scheduled", "in_progress", "final"]


class MarketResponse(BaseModel):
    """Response model for market list endpoints."""

    market_id: str
    game_id: str
    game: GameInfo | None = None
    market_type: Literal["spread", "moneyline", "total", "prop"]
    outcome_label: str
    line: float | None = None
    odds_decimal: float
    odds_american: int | None = None

    # Model outputs
    p_true: float = Field(ge=0, le=1, description="Calibrated model probability")
    p_market: float = Field(ge=0, le=1, description="Market-implied probability")
    raw_edge: float = Field(description="p_true - p_market")

    # Algorithm A outputs
    algo_a_value_score: float = Field(ge=0, le=100)
    algo_a_confidence: float
    algo_a_market_quality: float

    # Algorithm B outputs
    algo_b_value_score: float = Field(ge=0, le=100)
    algo_b_confidence: float
    algo_b_market_quality: float

    # Meta
    time_to_tip_minutes: int
    calc_time: datetime
    book: str | None = None

    model_config = {"from_attributes": True}


class MarketSummary(BaseModel):
    """Simplified market info for lists."""

    market_id: str
    market_type: str
    outcome_label: str
    value_score: float
    raw_edge: float
    time_to_tip_minutes: int

File: frontend/src/types/market.ts
export type MarketType = 'spread' | 'moneyline' | 'total' | 'prop';
export type GameStatus = 'scheduled' | 'in_progress' | 'final';
export type Algorithm = 'a' | 'b';

export interface Game {
  game_id: string;
  home_team: string;
  away_team: string;
  tip_time: string;
  status: GameStatus;
}

export interface Market {
  market_id: string;
  game_id: string;
  game?: Game;
  market_type: MarketType;
  outcome_label: string;
  line: number | null;
  odds_decimal: number;
  odds_american?: number;

  // Model outputs
  p_true: number;
  p_market: number;
  raw_edge: number;

  // Algorithm A outputs
  algo_a_value_score: number;
  algo_a_confidence: number;
  algo_a_market_quality: number;

  // Algorithm B outputs
  algo_b_value_score: number;
  algo_b_confidence: number;
  algo_b_market_quality: number;

  // Meta
  time_to_tip_minutes: number;
  calc_time: string;
  book?: string;
}

export interface MarketFilters {
  algorithm: Algorithm;
  marketType: MarketType | null;
  minValueScore: number;
  minConfidence: number;
}

export interface ConfidenceBreakdown {
  ensemble_agreement: number;
  calibration_reliability: number;
  injury_certainty: number;
  segment_reliability?: number;
  final_multiplier: number;
}

export interface MarketQualityBreakdown {
  liquidity_score: number;
  book_consensus: number;
  line_stability: number;
  final_multiplier: number;
}

export interface AlgorithmScore {
  algorithm: Algorithm;
  value_score: number;
  edge_score?: number;
  combined_edge?: number;
  confidence: ConfidenceBreakdown;
  market_quality: MarketQualityBreakdown;
}

export interface BetDetail extends Market {
  home_team: string;
  away_team: string;
  tip_time: string;
  edge_percentage: number;
  p_ensemble_mean: number;
  p_ensemble_std: number;
  algo_a: AlgorithmScore;
  algo_b: AlgorithmScore;
  active_algorithm: Algorithm;
  recommended_score: number;
}

export interface BetHistory {
  calc_time: string;
  p_true: number;
  p_market: number;
  raw_edge: number;
  algo_a_value_score: number;
  algo_b_value_score: number;
  odds_decimal: number;
  line?: number;
}
